Zone management cluster (SDK Impl, Camera-App Integration and Test scripts)

pidarped opened the pull request:
* Zone Management SDK Cluster logic
* CameraApp Integration
** Zone Manager delegate and relevant HAL interfaces
Cert test scripts:
* TC_ZONEMGMT_2.1 test script for testing attributes
* TC_ZONEMGMT_2.2 test script for testing CreateTwoDCartZone/RemoveZone commands
* TC_ZONEMGMT_2.3 test script for testing UpdateTwoDCartZone/RemoveZone commands
* TC_ZONEMGMT_2.4 test script for testing CreateorUpdateTriggers commands
* Unit tests for zone polygon self-intersection testing

#### Testing
* Commission camera-app with chip-tool/camera-controller
* Issue Read/Write operations on ZoneManagement cluster
** `./chip-tool zonemanagement read maxzones 1 1`
** `./chip-tool zonemanagement read max-user-defined-zones 1 1` 
* Issue Cread/Update/Remove Zone Management commands
** `./chip-tool zonemanagement create-two-dcartesian-zone '{"name": "zone1", "use": 0, "vertices": [{"x": 10, "y": 10}, {"x": 20, "y": 10}, {"x": 20, "y": 20}, {"x": 10, "y":20}], "color": "#00FFEE" }' 1 1` 
** `./chip-tool zonemanagement read zones 1 1`
**  `./chip-tool zonemanagement create-two-dcartesian-zone '{"name": "zone1", "use": 0, "vertices": [{"x": 10, "y": 10}, {"x": 20, "y": 20}, {"x": 10, "y": 20}, {"x": 20, "y":10}], "color": "#00FFEE" }' 1 1`  :  DynamicConstraintError due to self-intersection.
** `./chip-tool zonemanagement create-or-update-trigger '{"zoneID": 1, "initialDuration": 100, "augmentationDuration": 10, "maxDuration": 50, "blindDuration": 10, "sensitivity": 10}' 1 1` : Creates a trigger for ZoneId 1.
* Execute TC_ZONEMGMT_2_1.py for validating attributes
* Execute TC_ZONEMGMT_2_2.py for validating CreateTwoDCartesianZone command 
* Execute TC_ZONEMGMT_2_3.py for validating UpdateTwoDCartesianZone command 
* Execute TC_ZONEMGMT_2_4.py for validating CreateOrUpdateTrigger command

pidarped committed:
Initial zone management cluster skeleton

pidarped committed:
Update Zone Mgmt header with Delegate methods, Command and member
definitions.

pidarped committed:
Updates to Read/Write and InvokeCommand.

pidarped committed:
Restyle fix update

pidarped committed:
Documentation for Delegate APIs.

pidarped committed:
Zap template json and config-data updates.

pidarped committed:
Update camera-app.zap with zone-management cluster in EP1.

pidarped committed:
Fix typo in zcl json files

pidarped committed:
Run zap regen to include zone management cluster into camera-app.matter.

pidarped committed:
Add basic ZoneManagement delegate skeleton

pidarped committed:
Integrate initiation of Zone Management server in CameraApp

pidarped committed:
Fix build issues and add constraint error checks to CreateTwoDZone and
UpdateTwoDZone.

pidarped committed:
Add more Command handling checks and remove GetTwoDCartesianZone
command.

pidarped committed:
Define Storage structs for TwoDCartesianZone and ZoneInformation.

- Use the storage struct for the vector of stored zones.
- Modify Add and Update functions for Zones and Triggers.

pidarped committed:
Updates to providing the fixed parameters from the Camera HAL
and generate zoneIDs in the delegate implementation during TwoD zone
creation.

pidarped committed:
Re-run Zap-regen to resolve conflicts in generated code after rebase.

pidarped committed:
Updates to ZoneManagement Delegate and Camera-App integration

- Add HAL interfaces for trigger handling.
- Split CreateOrUpdateTrigger into separate Create and Update
  APIs when interfacing with the Delegate and HAL.
- Handle the sensitivity read/write parameter.

pidarped committed:
Set the camera-device in the ZoneManager.

pidarped committed:
Populate the TwoDCartesianZone field in ZoneInformation and fix a typo.

pidarped committed:
Remove unused commands in ZoneManagement cluster from camera-app.zap.

pidarped committed:
Add ZoneTriggered and ZoneStopped event generation APIs.

- Define OnEvent callbacks in Camera Device Interface that are implemented
  by the ZoneManager to call the SDK event generation APIs.

pidarped committed:
Add Duplicate Zone and Polygon self-intersection checks.

pidarped committed:
Ensure Mfg zones count is equal to MaxUserDefinedZones + MaxZones.

pidarped committed:
Add TC_ZONEMGMT_2.1 Test script for testing the Attributes.

pidarped committed:
Fix CI code-lint error in Python test script.

gmarcosb commented on src/app/clusters/zone-management-server/zone-management-server.cpp:
```diff
@@ -0,0 +1,767 @@
+    case Commands::CreateTwoDCartesianZone::Id:
+        ChipLogDetail(Zcl, "ZoneManagement[ep=%d]: Creating TwoDCartesian Zone", mEndpointId);
+
+        if (!HasFeature(Feature::kTwoDimensionalCartesianZone) || !HasFeature(Feature::kUserDefined))
```
Seems like you could put this in a simple define to make this a bit more readable, maybe in `CodeUtils.h` or `CommandHandlerInterface.h`?

```
#define ReturnIfNotSupportedCommand(handlerContext, expr)                                 \
    do                                                                                                                             \
    {                                                                                                                              \
        if (!(expr))                                                                                                               \
        {                                                                                                                          \
            handlerContext.mCommandHandler.AddStatus(handlerContext.mRequestPath, Status::UnsupportedCommand); \
            return; \
        }                                                                                                                          \
    } while (false)
```

Then here & throughout you wouldn't need the if/else, just:
```
ReturnIfNotSupportedCommand(handlerContext, HasFeature(Feature::kTwoDimensionalCartesianZone) && HasFeature(Feature::kUserDefined))
```

pidarped replied:
Yes. I think some of these can be cleaned up. Will use the VerifyOrReturn construct. Will update.

gmarcosb commented on src/app/clusters/zone-management-server/zone-management-server.h:
```diff
@@ -0,0 +1,356 @@
+    const TwoDCartesianVertexStruct mTwoDCartesianMax;
+    uint8_t mUserDefinedZonesCount = 0;
+
+    std::vector<ZoneInformationStorage> mZones;
```
Is it correct that `mZones` is stored in-memory & never loaded?

I.e. whose responsibility is it to persist zones?

pidarped replied:
This probably was a bad oversight. Both Zones and Triggers ought to be persisted I think. Will update here and post a Spec PR.

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-util.h:
```diff
@@ -0,0 +1,140 @@
+                    segments_are_adjacent = true;
+                }
+
+                // Also a logical adjacency check for sequence
```
Why?  If j = i+1, then p2 == q1, and why are we doing this extra check?

At the very least, this needs to document why such a check should be done.

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-util.h:
```diff
@@ -0,0 +1,140 @@
+            TwoDCartesianVertexStruct p1 = vertices[i];
+            TwoDCartesianVertexStruct q1 = vertices[(i + 1) % vertexCount];
+
+            for (size_t j = i + 1; j < vertexCount; ++j)
```
Consider starting at i+2, since you know i+1 will be skipped anyway.

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-util.h:
```diff
@@ -0,0 +1,140 @@
+    static int Orientation(const TwoDCartesianVertexStruct & p, const TwoDCartesianVertexStruct & q,
+                           const TwoDCartesianVertexStruct & r)
+    {
+        long long val = static_cast<long long>((q.y - p.y) * (r.x - q.x)) - static_cast<long long>((q.x - p.x) * (r.y - p.y));
```
This formula does not look right to me.  Where is it coming from?  This sort of thing should probably come with unit tests that would have ideally caught these problems.

It might help to explicitly document which vectors this is computing the cross-product of....

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-util.h:
```diff
@@ -0,0 +1,140 @@
+    // 0 --> p, q and r are collinear
+    // 1 --> Counterclockwise
+    // 2 --> Clockwise
+    static int Orientation(const TwoDCartesianVertexStruct & p, const TwoDCartesianVertexStruct & q,
```
Consider returning an enum for the orientation.

bzbarsky-apple commented on src/python_testing/TC_ZONEMGMT_2_1.py:
I did not review this file.

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-management-server.h:
```diff
@@ -0,0 +1,356 @@
+#include <app/AttributeAccessInterface.h>
+#include <app/CommandHandlerInterface.h>
+#include <app/StatusResponse.h>
+#include <app/reporting/reporting.h>
```
Where is reporting.h used in this header?

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-management-server.h:
```diff
@@ -0,0 +1,356 @@
+ *  Defines interfaces for implementing application-specific logic for various aspects of the ZoneManagement Cluster.
+ *  Specifically, it defines interfaces for the command handling and loading of the allocated streams.
+ */
+class ZoneMgmtDelegate
```
So why do we have ZoneManagement::ZoneMgmtDelegate?  Why not just ZoneManagement::Delegate?

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-management-server.h:
```diff
@@ -0,0 +1,356 @@
+    Protocols::InteractionModel::Status AddOrUpdateTrigger(const ZoneTriggerControlStruct & trigger);
+    Protocols::InteractionModel::Status RemoveTrigger(uint16_t zoneId);
+
+    // Send Zone events
```
```suggestion
    // Generate Zone events
```

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-management-server.cpp:
```diff
@@ -0,0 +1,767 @@
+#include <app/InteractionModelEngine.h>
+#include <app/clusters/zone-management-server/zone-management-server.h>
+#include <app/reporting/reporting.h>
+#include <app/util/attribute-storage.h>
```
Where does this file use attribute-storage.h?

pidarped replied:
Doesn't need to. Removed.

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-management-server.h:
```diff
@@ -0,0 +1,356 @@
+{
+};
+
+class ZoneMgmtServer : public CommandHandlerInterface, public AttributeAccessInterface
```
Again, why does the name repeat ZoneMgmt in it?

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-management-server.h:
```diff
@@ -0,0 +1,356 @@
+     * CommandHandler or AttributeHandler registration fails, else returns CHIP_NO_ERROR.
+     * This method also checks if the feature setting is valid, if invalid it will return CHIP_ERROR_INVALID_ARGUMENT.
+     */
+    CHIP_ERROR Init();
```
This seems to assume that any mfg zones are added before calling Init()?  Trying to understand how that Init check works if that's not the case.

pidarped replied:
On second thoughts, I felt it is best to remove this mfg-zone check in Init to allow for the flexibility of defining mfg zones later.

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-management-server.h:
```diff
@@ -0,0 +1,356 @@
+    /**
+     * @brief Initialise the Zone Management server instance.
+     * This function must be called after defining a ZoneMgmtServer class object.
+     * @return Returns an error if the given endpoint and cluster ID have not been enabled in zap or if the
```
This Init is not checking anything about "zap".

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-management-server.cpp:
```diff
@@ -0,0 +1,767 @@
+        return CHIP_IM_GLOBAL_STATUS(ConstraintError);
+    }
+
+    mDelegate.OnAttributeChanged(Attributes::Sensitivity::Id);
+    MatterReportingAttributeChangeCallback(ConcreteAttributePath(mEndpointId, ZoneManagement::Id, Attributes::Sensitivity::Id));
+
+    return CHIP_NO_ERROR;
```
This is all great, but where is the actual set happening?

Was this tested at all?  If not, why not?

I am stopping here in the .cpp.

pidarped replied:
It seems both the SetSensitivity() definition and call in Write() were added in commit `cb298cc8b53`, which was well before your review.

bzbarsky-apple replied:
Just for the record: the commit I reviewed did not have any changes to mSensitivity in Write() or anything called by Write().

bzbarsky-apple commented:
/gemini review

pidarped committed:
Address review comments.

pidarped committed:
Add initial APIs for persistence of Sensitivity, Zones, and Triggers.

pidarped committed:
Fix typo in camera-app.zap.

pidarped committed:
Miscellaneous minor fixes.
--Remove mfg-zones check in Init() function to allow for flexibility to
  add mfg zones later.

pidarped commented on src/app/clusters/zone-management-server/zone-management-server.cpp:
```diff
@@ -0,0 +1,767 @@
+
+CHIP_ERROR ZoneMgmtServer::SetSensitivity(uint8_t aSensitivity)
+{
+    VerifyOrReturnValue(aSensitivity != mSensitivity, CHIP_NO_ERROR);
```
Good catch!

pidarped commented on examples/camera-app/linux/include/clusters/zone-mgmt/zone-manager.h:
```diff
@@ -0,0 +1,101 @@
+
+    uint16_t GetNewZoneId()
+    {
+        // TODO: Replace with a better algo; Maybe use a std::set
```
The TODO comment is for employing a more efficient algo. This is O(N^2). But N is small.
idExists is reset to false everytime and should return if the newId is not found.

pidarped commented on src/app/clusters/zone-management-server/zone-management-server.cpp:
```diff
@@ -0,0 +1,767 @@
+
+    Status status = RemoveTrigger(zoneID);
+
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
```
It is not recursive. It is the Delegate call with the same name.

pidarped committed:
Add TC_ZONEMGMT_2.2 Test case for the CreateTwoDCartesianZone and
RemoveZone command.

yufengwangca commented on src/app/clusters/zone-management-server/zone-management-server.h:
```diff
@@ -0,0 +1,377 @@
+     * @brief Creates a Zone Management cluster instance. The Init() function needs to be called for this instance
+     * to be registered and called by the interaction model at the appropriate times.
+     *
+     * @param aDelegate                         A pointer to the delegate to be used by this server.
```
```suggestion
     * @param aDelegate                         A reference to the delegate to be used by this server.
```

yufengwangca commented on src/app/clusters/zone-management-server/zone-management-server.h:
```diff
@@ -0,0 +1,377 @@
+};
+
+enum class OptionalAttribute : uint32_t
+{
```
Enum OptionalAttribute is Empty, but you already use the version from zap in you code, 

CameraAvStreamManagement::OptionalAttribute::kImageFlipVertical)

why we still need to define our own version?

pidarped replied:
Yes, it is not needed for this cluster as there is no optional attribute. Had created the skeleton initially. Removed it.

yufengwangca commented on src/app/clusters/zone-management-server/zone-management-server.cpp:
```diff
@@ -0,0 +1,788 @@
+    {
+        RemoveTrigger(zoneID);
+    }
+    else
```
Duplicate Status Setting 

ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
if (status == Status::Success)
{
    RemoveTrigger(zoneID);  // <- Already done
}
else
{
    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);  // <-- Duplicate
    return;
}
ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Success); // <-- Also duplicate

Suggest:

if (status == Status::Success)
{
    RemoveTrigger(zoneID);
    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Success);
}
else
{
    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
}

yufengwangca commented on src/app/clusters/zone-management-server/zone-management-server.cpp:
```diff
@@ -0,0 +1,788 @@
+{
+    mZones.erase(
+        std::remove_if(mZones.begin(), mZones.end(), [&](const ZoneInformationStorage & zone) { return zone.zoneID == zoneId; }),
+        mZones.end());
```
Shall we return CHIP_ERROR_NOT_FOUND if the zone ID isn't found.

yufengwangca commented on src/app/clusters/zone-management-server/zone-management-server.cpp:
```diff
@@ -0,0 +1,788 @@
+
+    if (DoesZoneAlreadyExist(zoneToUpdate.use, twoDCartVertices))
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::AlreadyExists);
```
We need to check zoneID first

pidarped replied:
Added the zoneID check inside the DoesZoneAlreadyExist by passing the zoneID as a Nullable parameter.

yufengwangca commented on src/app/clusters/zone-management-server/zone-management-server.cpp:
```diff
@@ -0,0 +1,788 @@
+        return;
+    }
+
+    if (DoesZoneAlreadyExist(zoneToUpdate.use, twoDCartVertices))
```
```suggestion
    if (zone.zoneID != zoneID && DoesZoneAlreadyExist(zoneToUpdate.use, twoDCartVertices))
```

yufengwangca commented on src/app/clusters/zone-management-server/zone-management-server.cpp:
```diff
@@ -0,0 +1,788 @@
+    case Sensitivity::Id:
+        VerifyOrReturnError(!HasFeature(Feature::kPerZoneSensitivity), CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute),
+                            ChipLogError(Zcl,
+                                         "ZoneManagement[ep=%d]: can not get Sensitivity, PerZoneSensitivity feature is supported",
```
```suggestion
                                         "ZoneManagement[ep=%d]: Cannot read Sensitivity: PerZoneSensitivity feature is enabled",
```

pidarped committed:
Address review comments

pidarped commented on src/app/clusters/zone-management-server/zone-util.h:
```diff
@@ -0,0 +1,149 @@
+/*
+ * Utility Class providing functionality around various checks required for
+ * zones, such as self-intersection, duplicates, etc.
```
Added.

pidarped commented on src/app/clusters/zone-management-server/zone-management-server.h:
```diff
@@ -0,0 +1,377 @@
+    virtual CHIP_ERROR LoadTriggers(std::vector<ZoneTriggerControlStruct> & aTriggers) = 0;
+
+private:
+    friend class ZoneMgmtServer;
```
The friend allows the ZoneMgmtServer to access the Delegate privates, in this case the private Setter to populate the pointer in the ZoneMgmtServer ctor.

pidarped committed:
Add TC_ZONEMGMT_2.3 Test case for UpdateTwoDCartesianZone command.

pidarped committed:
Add TC_ZONEMGMT_2.4 Test case for CreateOrUpdateTrigger, RemoveTrigger and RemoveZone commands.

pidarped committed:
Modify TC_ZONEMGMT_2_3 to create a new TwoDCartZone before sending the
Update command for the duplicate zone that returns AlreadyExists status.

This is to handle the recent TestPlan and Spec changes made around
duplicate detection during UpdateTwoDCartZone command.

pidarped committed:
Add CameraAppCommandDelegate for command driven event generation.

These would be leveraged by the Test Harness scripts to trigger events
from the Camera App.

pidarped committed:
Fix linter errors

pidarped committed:
Add ZoneTriggered event generation from Test script

joyoungjun commented on src/python_testing/TC_ZONEMGMT_2_1.py:
```diff
@@ -0,0 +1,174 @@
+                endpoint=endpoint, cluster=Clusters.CameraAvStreamManagement, attribute=Clusters.CameraAvStreamManagement.Attributes.VideoSensorParams
+            )
+            logger.info(f"Rx'd VideoSensorParams: {videoSensorParams}")
+            asserts.assert_true(twoDCartesianMax.x < videoSensorParams.sensorWidth and twoDCartesianMax.y <
+                                videoSensorParams.sensorHeight, "TwoDCartesianMax should be within the VideoSensorParams dimensions")
```
```suggestion
            asserts.assert_true(twoDCartesianMax.x == videoSensorParams.sensorWidth -1 and twoDCartesianMax.y  == videoSensorParams.sensorHeight - 1, "TwoDCartesianMax should be within the VideoSensorParams dimensions")
```

joyoungjun commented on src/python_testing/TC_ZONEMGMT_2_2.py:
```diff
@@ -0,0 +1,390 @@
+    def steps_TC_ZONEMGMT_2_2(self) -> list[TestStep]:
+        return [
+            TestStep("1", "Commissioning, already done", is_commissioning=True),
+            TestStep("2a", "If DUT supports TwoDCartesianZone and User defined zones, TH sends CreateTwoDCartesianZone command with",
+                     " valid parameters",
+                     "Verify that the DUT response contains a new zoneId and the corresponding zone information matches."),
+            TestStep("2b", "TH sends RemoveZone command with ZoneID set to the created ZoneID in step 2a ",
```
```suggestion
            TestStep("2", "If DUT supports TwoDCartesianZone and User defined zones, TH sends CreateTwoDCartesianZone command with",
                     " valid parameters",
                     "Verify that the DUT response contains a new zoneId and the corresponding zone information matches."),
            TestStep("2a", "TH sends RemoveZone command with ZoneID set to the created ZoneID in step 2 ",
```

joyoungjun commented on src/python_testing/TC_ZONEMGMT_2_2.py:
```diff
@@ -0,0 +1,390 @@
+        self.focusZonesSupported = aFeatureMap & cluster.Bitmaps.Feature.kFocusZones
+        self.perZoneSenseSupported = aFeatureMap & cluster.Bitmaps.Feature.kPerZoneSensitivity
+
+        self.step("2a")
```
```suggestion
        self.step("2")
```

joyoungjun commented on src/python_testing/TC_ZONEMGMT_2_2.py:
```diff
@@ -0,0 +1,390 @@
+            asserts.assert_equal(matchingZone.zoneSource, enums.ZoneSourceEnum.kUser,
+                                 "ZoneSource of created Zone is not of type User")
+
+            self.step("2b")
```
```suggestion
            self.step("2a")
```

joyoungjun commented on src/python_testing/TC_ZONEMGMT_2_3.py:
```diff
@@ -0,0 +1,414 @@
+
+            matchingZone = next(
+                (z for z in zonesAfterUpdate if z.zoneID == zoneID1), None)
+            asserts.assert_is_not_none(matchingZone, "fZone with {zoneID1} not found")
+            asserts.assert_equal(matchingZone.zoneType, enums.ZoneTypeEnum.kTwoDCARTZone,
+                                 "ZoneType of updated Zone is not of type TwoDCartZone")
+            asserts.assert_equal(matchingZone.zoneSource, enums.ZoneSourceEnum.kUser,
+                                 "ZoneSource of updated Zone is not of type User")
+            asserts.assert_equal(matchingZone.twoDCartesianZone.name, "Zone1_Updated",
+                                 "Name of updated Zone does not match")
```
It seems necessary to validate whether ZoneVertices has been properly updated.

joyoungjun commented on src/python_testing/TC_ZONEMGMT_2_3.py:
```diff
@@ -0,0 +1,414 @@
+                    "Unexpected error returned expecting NOT_FOUND due to new zone id being used",
+                )
+                pass
+
```
```suggestion

            self.step("4")
```

joyoungjun commented on src/python_testing/TC_ZONEMGMT_2_4.py:
```diff
@@ -0,0 +1,419 @@
+
+            matchingZone = next(
+                (z for z in zonesAfterCreate if z.zoneID == zoneID1), None)
+            asserts.assert_is_not_none(matchingZone, "fZone with {zoneID1} not found")
+            asserts.assert_equal(matchingZone.zoneType, enums.ZoneTypeEnum.kTwoDCARTZone,
+                                 "ZoneType of created Zone is not of type TwoDCartZone")
+            asserts.assert_equal(matchingZone.zoneSource, enums.ZoneSourceEnum.kUser,
+                                 "ZoneSource of created Zone is not of type User")
```
It seems necessary to validate whether ZoneVertices has been properly created.

pidarped replied:
Added assert checks for verifying if the zone vertices match.

joyoungjun commented on src/python_testing/TC_ZONEMGMT_2_4.py:
```diff
@@ -0,0 +1,419 @@
+
+            matchingZone = next(
+                (z for z in zonesAfterCreate if z.zoneID == zoneID4), None)
+            asserts.assert_is_not_none(matchingZone, "fZone with {zoneID4} not found")
+            asserts.assert_equal(matchingZone.zoneType, enums.ZoneTypeEnum.kTwoDCARTZone,
+                                 "ZoneType of created Zone is not of type TwoDCartZone")
+            asserts.assert_equal(matchingZone.zoneSource, enums.ZoneSourceEnum.kUser,
+                                 "ZoneSource of created Zone is not of type User")
```
It seems necessary to validate whether ZoneVertices has been properly created.

joyoungjun commented on src/python_testing/TC_ZONEMGMT_2_4.py:
```diff
@@ -0,0 +1,419 @@
+        asserts.assert_equal(matchingTrigger.maxDuration, 50,
+                             "MaxDuration of created Trigger does not match")
+        asserts.assert_equal(matchingTrigger.blindDuration, 5,
+                             "BlindDuration of created Trigger does not match")
```
If the DUT supports PERZONESENS, it seems necessary to verify the Sensitivity value as well.

pidarped replied:
Added assert checks for verifying if the zone vertices match.

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-management-server.cpp:
```diff
@@ -0,0 +1,790 @@
+void ZoneMgmtServer::HandleCreateOrUpdateTrigger(HandlerContext & ctx,
+                                                 const Commands::CreateOrUpdateTrigger::DecodableType & commandData)
+{
+    ZoneTriggerControlStruct trigger = commandData.trigger;
```
Where is the constraint validation on the fields of this struct?

pidarped replied:
Missed adding them. Thanks. Fixed.

pidarped committed:
Add zone trigger timer and associated logic for exercising ZoneTriggered
and ZoneStopped events.

Add sub-tests for emitting ZoneStopped event for both ActionStopped and
Timeout scenarios.

pidarped committed:
Apply suggestions from code review

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>
Co-authored-by: joyoungjun <youngjun4724.jo@lge.com>

pidarped committed:
Address review comments

-Run zap_regen for incorporating updated zap json files

pidarped commented on src/app/clusters/zone-management-server/zone-management-server.cpp:
```diff
@@ -0,0 +1,790 @@
+        ZoneInformationStorage zoneInfo;
+        zoneInfo.Set(zoneID, ZoneTypeEnum::kTwoDCARTZone, ZoneSourceEnum::kUser, MakeOptional(twoDCartZoneStorage));
+
+        UpdateZone(zoneID, zoneInfo);
```
Have kept the return value in the API for completeness, so that it does not need to be changed later if required.

pidarped commented on src/app/zap-templates/zcl/zcl-with-test-extensions.json:
```diff
@@ -793,12 +793,18 @@
+            "MaxZones",
+            "SensitivityMax",
+            "Sensitivity",
+            "TwoDCartesianMax"
```
Added.

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-util.h:
```diff
@@ -0,0 +1,167 @@
+    // and hence the orientation of the 3 coordinates. A value of 0 indicates that the points
+    // are collinear.
+    // Using the right-hand rule Z > 0 means pointing upward from the 2D
+    // plane(counter-clockwise rotation) and Z < 0 means pointing downward into
```
> The vectors being considered are pq and pr and the rotation is from pq to pr

My point s that this should be _documented_.

pidarped replied:
Was already documented above the CrossProduct() function.

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-util.h:
```diff
@@ -0,0 +1,167 @@
+                TwoDCartesianVertexStruct p2 = vertices[j];
+                TwoDCartesianVertexStruct q2 = vertices[(j + 1) % vertexCount];
+
+                // Skip segments that share an endpoint (i.e., adjacent segments)
```
OK, so still trying to understand the intent here: say our vertex sequence is: `(0,0), (1,0), (2,0), (2,1), (1,0), (0,1)`.  Should that be considered self-intersecting?  Does this code consider it self-intersecting?

What about: `(0,0), (2,0), (2,1), (2,0), (1,0)`?  Same questions.

What about: `(0,0), (1,0), (0,0), (0,1)`?  Here I am quite sure the code as written does not consider this to be self-intersecting, but does that match the expected behavior?

pidarped replied:
Ok. So, looking again, it seems the above set of vertices should be disallowed because they do not form a _simple polygon_ , and the reason is because they have at least one repeated vertex in the sequence. The zones are expected to be simple polygons that form a single, unbroken loop without any edges crossing over each other, and without any _holes_ formed by repeating intermediate vertices.
So, towards that, I think I need to add a check upfront for repeated vertices. The status code for that should be `InvalidCommand`, right? Or should it be `ConstraintError`? I am thinking, the former.

bzbarsky-apple replied:
OK, so if we disallow repeated vertices, what about: `(0,0), (2,0), (1,0)`?  I believe the current code will not treat that as "self-intersecting".  Should it?

bzbarsky-apple replied:
No, I think the general problem remains, because you could have a degenerate triangle like this inside any polygon with any vertex count, no?

I would strongly suggest thinking about a good way to do the self-intersection check in a way that does not rely on too much special-casing.

pidarped replied:
>No, I think the general problem remains, because you could have a degenerate triangle like this inside any polygon with any vertex count, no?

Yes, potentially. But that should get filtered by the duplicate vertex check, isn't it?

pidarped replied:
Based on the proof provided, it is sufficient to check non-adjacent edges for all polygons > 3 vertices. So, it has the 3-vertex-only case treated specially up-front for collinearity.

pidarped committed:
Fix the handling of zones with duplicate vertices and
other degenerate cases.

Add tests for several different self-intersecting zones
during Creation and Update of zones.

pidarped committed:
Run zap regen to resolve conflicts.

pidarped committed:
Outline Self-Intersection algorithm and provide CI param for app-pipe.

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-util.h:
```diff
@@ -0,0 +1,250 @@
+    }
+
+    /**
+     * Algorithm for Checking Self-Intersection of a Zone
```
This needs some proof-like thing that this algorithm actually works correctly.

pidarped replied:
Sure. Will provide a proof here. 
Just to highlight the 2 changes that were added to the initial algorithm:
1) Filtering polygons with repeated vertices as a basic prerequisite for _simple_ polygons. This handles the vast majority of degenerate cases.
2) The special case of allowing self-intersection checks on edges of the 3-vertex polygon(with all edges adjacent to each other) which is required because of the optimized logic of checking for self-intersection _only_ among non-adjacent edges.

pidarped replied:
Added Proof of Correctness below the Algorithm outline.

pidarped replied:
Modified the algorithm slightly to have the 3-vertex-only case handled up-front and flagging it as self-intersecting if they are collinear. The algorithm of picking non-adjacent edges, thus, runs on all polygons > 3 vertices.

pidarped committed:
Add a proof of correctness for the zone self-intersection checking
algorithm.

pidarped committed:
Add whitespace before '(' in some comment lines.

pidarped committed:
Add unit tests for zone self-intersections.

pidarped committed:
Augment the algorithm correctness by adding a proof of a sub-claim.

pidarped committed:
Fix linter to have the new renamed file.

pidarped committed:
Add more unit tests and flag the case of only 3 collinear vertices up-front.

-Rename CrossProduct function as Orientation and use vertex indices for
 checking adjacency.

pidarped commented on examples/camera-app/camera-common/include/camera-app.h:
```diff
@@ -47,6 +47,7 @@ class CameraApp
     std::unique_ptr<chip::app::Clusters::CameraAvStreamManagement::CameraAVStreamMgmtServer> mAVStreamMgmtServerPtr;
     std::unique_ptr<chip::app::Clusters::CameraAvSettingsUserLevelManagement::CameraAvSettingsUserLevelMgmtServer>
         mAVSettingsUserLevelMgmtServerPtr;
+    std::unique_ptr<chip::app::Clusters::ZoneManagement::ZoneMgmtServer> mZoneMgmtServerPtr;
```
Keeping it as-is for now, because none of the camera clusters follow the model of naming the server class `Instance`. Maybe we can revisit renaming them consistently. I also felt that the repetition is only felt in the header file but the vast majority of cpp file usage can use the single name without needing the `::` separated combined name.

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-geometry.h:
```diff
@@ -0,0 +1,335 @@
+        return q.x <= std::max(p.x, r.x) && q.x >= std::min(p.x, r.x) && q.y <= std::max(p.y, r.y) && q.y >= std::min(p.y, r.y);
+    }
+
+    // Helper function: Determine the orientation of the ordered triplet (p, q, r) of vertices.
+    // The purpose of this function is to determine if 'r' is to the left, right
+    // of the vector 'pq', or is collinear with it.
```
```suggestion
    // Helper function: Determine the orientation of the point r with respect to ray pq: in
    // the left half-plane when facing in the direction of the ray, in the right half-plane,
    // or collinear with the ray.
```

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-geometry.h:
```diff
@@ -0,0 +1,335 @@
+    // vectors pq and pr to find the direction of the Z axis. The rotation
+    // considered is from vector pq to pr.
+    // The sign of the determinant is used to infer the direction of the cross-product vector
+    // and hence the orientation of the 3 coordinates. A value of 0 indicates that the points
```
```suggestion
    // and hence the orientation of the point with respect to the ray. A value of 0 indicates that the points
```

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-geometry.h:
```diff
@@ -0,0 +1,335 @@
+    static OrientationEnum Orientation(const TwoDCartesianVertexStruct & p, const TwoDCartesianVertexStruct & q,
+                                       const TwoDCartesianVertexStruct & r)
+    {
+        long long val = static_cast<long long>((q.x - p.x) * (r.y - p.y)) - static_cast<long long>((q.y - p.y) * (r.x - p.x));
```
What are the static_casts doing here?  They don't make sense.... If the worry is possible overflow on the multiplication, then you need to cast _before_ multiplying, no?

bzbarsky-apple commented on src/app/clusters/zone-management-server/zone-management-server.cpp:
```diff
@@ -0,0 +1,826 @@
+
+    VerifyOrReturnError(trigger.maxDuration >= trigger.initialDuration, Status::ConstraintError);
+
+    if (HasFeature(Feature::kPerZoneSensitivity))
+    {
+        VerifyOrReturnError(trigger.sensitivity.Value() >= 1 && trigger.sensitivity.Value() <= mSensitivityMax,
+                            Status::ConstraintError);
+    }
```
This code does not match the spec, and will crash on one variety of invalid input (i.e. when the sensitivity is not present but the feature bit is set).

pidarped replied:
Yes, missed adding the absence/presence validation of that field in the command args. Added that.

pidarped committed:
Apply suggestions from code review

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

pidarped committed:
Address review comments

-Move zone-geometry.h from app-config-dependent-sources.gni to BUILD.gn.

-Minor editorial changes

pidarped commented on examples/camera-app/linux/src/clusters/zone-mgmt/zone-manager.cpp:
```diff
@@ -0,0 +1,257 @@
+        trigCtxtIter->timeSinceInitialTrigger += kTimerPeriod;
+
+        // ChipLogProgress(Camera, "Time since initial trigger = %u, triggerDetectedDuration = %u",
+        // trigCtxtIter->timeSinceInitialTrigger, trigCtxtIter->triggerDetectedDuration);
```
Removed. Remnant from past debugging I think.

pidarped commented on src/app/clusters/zone-management-server/app_config_dependent_sources.gni:
```diff
@@ -0,0 +1,18 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+app_config_dependent_sources = [
+  "zone-geometry.h",
```
Moved to BUILD.gn.

bzbarsky-apple commented on src/app/clusters/zone-management-server/app_config_dependent_sources.cmake:
```diff
@@ -0,0 +1,21 @@
+  PRIVATE
+    "${CLUSTER_DIR}/zone-management-server.cpp"
+    "${CLUSTER_DIR}/zone-management-server.h"
+)
```
Unfortunately, the `zone-geometry.h` bit _does_ need to stay in the cmake for now...  We don't have any other build definitions with cmake, sadly.

pidarped replied:
Moved it back.
